data_safe_haven/infrastructure/programs/sre/firewall.py

```diff
@@ -108,206 +111,236 @@ def __init__(
+                            network.AzureFirewallApplicationRuleProtocolArgs(
+                                port=int(Ports.HTTP),
+                                protocol_type=network.AzureFirewallApplicationRuleProtocolType.HTTP,
+                            ),
+                            network.AzureFirewallApplicationRuleProtocolArgs(
+                                port=int(Ports.HTTPS),
+                                protocol_type=network.AzureFirewallApplicationRuleProtocolType.HTTPS,
+                            ),
+                        ],
+                        source_addresses=props.subnet_workspaces_prefixes,
+                        target_fqdns=ForbiddenDomains.UBUNTU_SNAPCRAFT,
+                    ),
+                ],
+            ),
+        ]
+
+        network_rule_collections: list[
+            network.AzureFirewallNetworkRuleCollectionArgs
+        ] = []
+
+        if props.allow_workspace_internet:
+            application_rule_collections = []
+            network_rule_collections.append(
+                network.AzureFirewallNetworkRuleCollectionArgs(
                     action=network.AzureFirewallRCActionArgs(
-                        type=network.AzureFirewallRCActionType.DENY
+                        type=network.AzureFirewallRCActionType.ALLOW
                     ),
-                    name="workspaces-deny",
-                    priority=FirewallPriorities.SRE_WORKSPACES_DENY,
+                    name="workspaces-all-allow",  # TODO: Fix other names.
+                    priority=FirewallPriorities.SRE_WORKSPACES,
                     rules=[
-                        network.AzureFirewallApplicationRuleArgs(
-                            description="Deny external Ubuntu Snap Store upload and login access",
-                            name="DenyUbuntuSnapcraft",
-                            protocols=[
-                                network.AzureFirewallApplicationRuleProtocolArgs(
-                                    port=int(Ports.HTTP),
-                                    protocol_type=network.AzureFirewallApplicationRuleProtocolType.HTTP,
-                                ),
-                                network.AzureFirewallApplicationRuleProtocolArgs(
-                                    port=int(Ports.HTTPS),
-                                    protocol_type=network.AzureFirewallApplicationRuleProtocolType.HTTPS,
-                                ),
-                            ],
+                        network.AzureFirewallNetworkRuleArgs(
+                            description="Enables internet access to workspaces.",
+                            destination_addresses=["*"],
+                            destination_ports=["*"],
+                            name="allow-internet-access",
+                            protocols=[network.AzureFirewallNetworkRuleProtocol.ANY],
                             source_addresses=props.subnet_workspaces_prefixes,
```

Comment:
I've also tried fixing this locally (with option 2 from above) but still can't browse to anything from within the SRE.

Internet access is possible: I can `ftp` to `ftp.ubuntu.com` from a workspace.

So I think we need to also look at the DNS.
We have a local container that we use for DNS that (among other things) blocks most FQDNs from resolving.

Reply:
Hi @craddm , thanks for checking into this. Just to check my understanding, when you tested Option 2 did you have all the application rules enabled + the new network rule? Also, changing the DNS container means modifying `data_safe_haven/resources/dns_server/AdGuardHome.mustache.yaml`?

Reply:
You might need to slightly change `AdGuardHome.mustache.yaml`. Currently it forbids everything then allows everything from a specified allowlist. We might need to make the first part configurable as otherwise the rules will be "deny everything" and "allow everything" and I have no idea what that will do!